libzecale/circuits/aggregator_circuit_wrapper.tcc

```diff
@@ -13,113 +15,164 @@ namespace libzecale
 {
 
 template<
-    typename nppT,
     typename wppT,
-    typename nsnarkT,
-    typename wverifierT,
+    typename wsnarkT,
+    typename nverifierT,
+    typename hashT,
     size_t NumProofs>
-aggregator_circuit_wrapper<nppT, wppT, nsnarkT, wverifierT, NumProofs>::
+aggregator_circuit_wrapper<wppT, wsnarkT, nverifierT, hashT, NumProofs>::
     aggregator_circuit_wrapper(const size_t inputs_per_nested_proof)
-    : num_inputs_per_nested_proof(inputs_per_nested_proof)
+    : _num_inputs_per_nested_proof(inputs_per_nested_proof), _pb()
 {
+    // The order of allocation here is important as it determines which inputs
+    // are primary.
+
+    // Input for hash of nested verification key.
+    _nested_vk_hash.allocate(_pb, FMT("", "_nested_vk_hash"));
+
+    // For each proof in a batch, allocate primary inputs and results. These
+    // are the primary inputs. Note: both inputs and results will be
+    // populated by the aggregator gadget.
+    for (size_t i = 0; i < NumProofs; i++) {
+        _nested_primary_inputs[i].allocate(
```

Comment:
I moved the allocation and ordering of inputs out of the gadget. A "circuit wrapper" seemed much more appropriate. My thinking was:
- Having the gadget completely define how the inputs are allocated makes it slightly less usable (the user of the gadget should be able to determine whether each input is primary / auxiliary and allocate accordingly, and whether / how it might interact with other gadget in the circuit).
- The gadget was already pretty large and complex. Adding the key verification required adding more complexity to that gadget.
- Having the caller determine the inputs / outputs is consistent with other gadget objects (presumably for the reasons above).

Reply:
Hum.. The role of the wrapper was mostly to pass this "fresh" protoboard so far.

> Having the gadget completely define how the inputs are allocated makes it slightly less usable (the user of the gadget should be able to determine whether each input is primary / auxiliary and allocate accordingly, and whether / how it might interact with other gadget in the circuit).

The definition of the instance/witness is tightly related to the definition of the underlying circuit, so being able to `"to determine whether each input is primary / auxiliary and allocate accordingly"` does not seem - to me - to be a very valuable thing here. The circuit is the "algebraic embodiment" of statements of the form `x \in L`, where `L = {x | \exists w s.t. (x,w) \in R}`, so by changing how `x` and `w` are treated we change the language. While this allows to have "more generic" gadgets that could be re-used to represent different languages, I'd be careful here. I'd balance the advantages of this new syntax with the fact that things here aren't consistent with how we define circuits in Zeth and libsnark. I do find the cost of breaking this "consistency" across our projects to be quite high. For such code (the circuit code is pretty critical to get right) I tend to value consistency and "systematism" quite a lot as it makes reviews easier and makes errors/inconsistencies much easier to spot.

Reply:
Perhaps I'm missing something.  No other gadgets in libsnark determine how their inputs are allocated, so I can only see that this change fixes the previous inconsistency.  Gadgets should be reusable blocks of functionality, which is exactly why they do not allocate their inputs.

> The circuit is the "algebraic embodiment" of statements of the form x \in L, where L = {x | \exists w s.t. (x,w) \in R}, so by changing how x and w are treated we change the language.

This is true, and to me this is more justification for why the "circuit" class, not individual gadgets, should determine the allocation of inputs and in turn the structure of $L$ in your statment.  Note that your statement about L here does not reference "gadgets".  They are just building blocks of the circuit.

Just to illustrate, the changes here allow us to create different *circuits* that reuse the `aggregator_gadget` (hypotheitcally we could create a circuit that verifies 2 batches, or that does not verify the vk_hash, etc.).  Before these changes, the *gadget* defines quite a lot about the *circuit*.  This seems inconsistent with all the other libsnark gadgets that I have seen, which do not allocate their inputs and therefore leave the circuit structure, and in turn $L$, to the caller.

I'm probably missing something, but I'm still failing to see the justification for having a *gadget* dictate the circuit in this way.

Reply:
> They are just building blocks of the circuit.

A gadget is a circuit. Each gadget/circuit can be composed with others to form a bigger circuit (there exists a composition rule - which is why we need the "ONE" variable).
The main point is: yes I do see your point of keeping all gadgets generic - which is why the libsnark library has defined gadgets in that way in the first place, but the whole circuit (e.g. the `aggregator` circuit, or `joinsplit` circuit) is fixed and embodies **one specific language** (see my point above, i.e. the `aggregator` circuit -> "Zecale language" and `joinsplit` circuit -> "Zeth language"). That's why I don't see why we need to make such circuit generic, they just don't need to be - won't be re-used anywhere (and this is why the `joinsplit` circuit is implemented the way it is in Zeth). Now sure, we can make the whole circuit more generic and use the wrapper as you did, but to my points above, this breaks the consistency with the Zeth code base - where we don't do that.
That's not necessarily a major deal, but I would rather try to stick to a systematic approach across our projects to make sure our various circuits are consistently written. So if this is the way to go, then I'd be in favor to also carry out this refactor on Zeth, and extend the `circuit-wrapper` there.
As to the inconsistency with libsnark, I am mostly referring to the fact `libsnark` manipulates smart ptrs to gagdets to form the circuits, while here (in Zecale), some of these smart ptrs have been switched to actual class instances.
E.g. see some of the `fields` gadgets like
- https://github.com/clearmatics/libsnark/blob/master/libsnark/gadgetlib1/gadgets/fields/fp6_gadgets.tcc#L87-L142 in libsnark and https://github.com/clearmatics/zecale/blob/key-verification/libzecale/circuits/fields/fp6_3over2_gadgets.tcc#L173-L228 in Zecale

Again, I don't see any major deal here, but to my point above, this is another point of divergence with some existing code. So that would be nice IMO to try and be as consistent as possible with how we define gadgets in all our projects.

Reply:
> A gadget is a circuit. Each gadget/circuit can be composed with others to form a bigger circuit (there exists a composition rule - which is why we need the "ONE" variable).
> The main point is: yes I do see your point of keeping all gadgets generic - which is why the libsnark library has defined gadgets in that way in the first place, but the whole circuit (e.g. the aggregator circuit, or joinsplit circuit) is fixed and embodies one specific language (see my point above, i.e. the aggregator circuit -> "Zecale language" and joinsplit circuit -> "Zeth language").

I think there is / should be a bit more of a difference between "circuit" and "gadget".  You state above that that:

> The circuit is the "algebraic embodiment" of statements of the form x \in L, where L = {x | \exists w s.t. (x,w) \in R}, so by changing how x and w are treated we change the language.

That makes sense to me, but if a circuit defines the language then it is not really composible (in fact, this is exactly what motivated this change - I was unable to compose the batch verification "gadget" with the VK hashing "gadget" exactly because of this).  Gadgets, on the other hand, do *not* seem to define the language.  Either in libsnark or in zeth, *exept* for the joinsplit gadget.  I may have missed one, but I checked a couple of gadgets in the zeth codebase and they seem follow the libsnark convention where the caller passes in the input and output variables.  (Of course, most gadgets have some natural configuration of public/aux inputs, but it doesn't seem to be the case that they define this.  We may imagine gadgets that can be used in multiple configurations.)

I don't know the joinsplit gadget very well, but that seems to be the only place where the consisency is broken.  The convention itself seems pretty valuable, and so I think the correct approach here is to use your definition of "circuit" (i.e. complete definition including the language, and consequently NOT composable) and the definition of "gadget" implied by gadgets except `joinsplit_gadget`, (i.e. a composable part of a circuit).  The zeth codebase even has both a joinsplit "gadget" and "circuit wrapper" class, so it should be pretty easy to make it follow this convention.

> That's why I don't see why we need to make such circuit generic, they just don't need to be - won't be re-used anywhere (and this is why the joinsplit circuit is implemented the way it is in Zeth).

I'd say the gadget should be generic (otherwise it's not a gadget), and the circuit by definition cannot be generic.  Hence the `aggregator_circuit_wrapper` is completely un-generic, just setting up the inputs and the top-level gadget composition.  This is partly for consistency, as above, and so that we don't have single classes with too many concerns.  Proof aggregation is arguably a separate concern to hashing of the verifcation key, hence another "circuit" may want the verification key to be constant or hasing of the nested inputs, say (by rearranging the inputs and using some gadgets but not others),.  So I'd say there is a strong engineering argument for this approach.

> Now sure, we can make the whole circuit more generic and use the wrapper as you did, but to my points above, this breaks the consistency with the Zeth code base - where we don't do that.
> That's not necessarily a major deal, but I would rather try to stick to a systematic approach across our projects to make sure our various circuits are consistently written. So if this is the way to go, then I'd be in favor to also carry out this refactor on Zeth, and extend the circuit-wrapper there.

Right.  I think it makes sense to make `joinsplit_gadget` and `circuit_wrapper` in Zeth fit with the above.  For consistency with other code, and for separation of concerns.

> As to the inconsistency with libsnark, I am mostly referring to the fact libsnark manipulates smart ptrs to gagdets to form the circuits, while here (in Zecale), some of these smart ptrs have been switched to actual class instances.

In my mind, that's much more about the implementation detail and might require quite a bit more thought.  TBH I wonder if a lot of smart pointer usage in libsnark follows from a clear convention, or if it's more of a accidental side-effect of the gadget interface.  We may want to resolve these conceptual inconsistencies with libsnark, like "gadget", before tackling the implementation details.

Reply:
From [Madars' thesis on libsnark](https://madars.org/phd-thesis/Madars-Virza-thesis-20170831.pdf)
(page 26)
> The design of gadgetlib1. 
To obtain maximally efficient low-level representations of interesting NP relations, we built a library for wiring non-deterministic circuit “gadgets”,emphasizing modularity and non-determinism. The gadgetlib1 library is a suite of optimized subcircuit “gadgets” together with means of effectively composing, testing, and debugging them. Each gadget is a pair of two algorithms:instance map, which generates the circuit wiring, and witness map, which computes the wire values of the gadget, including assignments to non-deterministically set wires. It is exactly the use of non-determinism that sets apart gadget from a traditional arithmetic circuit, where all wire values can be computed directly from the circuit wiring.

Then page 28:
> In a sense, the gadget writing can be compared to directly writing “circuit silicon”, or constructing ASICs

All of this is in line with the composition rule defined in [GGPR](https://eprint.iacr.org/2012/215.pdf) (see section 7) as per my point above 
> A gadget is a circuit. Each gadget/circuit can be composed with others to form a bigger circuit (there exists a composition rule - which is why we need the "ONE" variable).

So yes sure a gadget is composable (and thus reusable) - that's the goal -, but it can be used in a standalone fashion, e.g. one may want to use: https://github.com/clearmatics/zeth/blob/master/libzeth/circuits/sha256/sha256_ethereum.hpp#L44-L49 as "the circuit" in a protocol to represent a language of interest (e.g. I have a protocol where I prove knowledge of the pre-image of a hash for instance). In this case, there is no wiring that needs to be carried out to route the output wire of a gadget to the inputs of another gadget - I just need to call my gadget on an input ("my gadget is my circuit"). In this case `sha256_ethereum` defines the language L, containing all pre-images of the hash, i.e. `L = {sha256_ethereum(w), \forall w \in {0,1}^*}`.

> In my mind, that's much more about the implementation detail and might require quite a bit more thought. TBH I wonder if a lot of smart pointer usage in libsnark follows from a clear convention, or if it's more of a accidental side-effect of the gadget interface. We may want to resolve these conceptual inconsistencies with libsnark, like "gadget", before tackling the implementation details.

Yes I don't disagree :) that's just another source of inconsistency I noticed across the projects, and that's why I think we need to make sure we write circuit code consistently to make sure we remove frictions (while writing code and reviewing code) and have a "systematic approach" to writing circuits, which IMHO makes everyone's life easier :)

Reply:
Perhaps this is just semantic detail, but I would say that the language is defined in a large part by variable allocations and a call to `protoboard::set_input_sizes`, all made outside of the `sha256_ethereum` gadget.

But it sounds like there is no issue with the main points re: the roles of gadget vs circuit wrapper.  I've created clearmatics/zeth#290 to deal with making the zeth side consistent.  If there are issues let me know and we can remove or change it.

Reply:
> But it sounds like there is no issue with the main points re: the roles of gadget vs circuit wrapper. I've created clearmatics/zeth#290 to deal with making the zeth side consistent. If there are issues let me know and we can remove or change it.

Great, sounds good to me. 
If we follow this approach, what about renaming `aggregator_circuit_wrapper.hpp ` as simply `aggregator_circuit` then?

Reply:
> what about renaming aggregator_circuit_wrapper as simply aggregator_circuit then?

Good idea.  Renamed in latest commit.
Thanks.